### PR TITLE
 Defining ImageSet Union

### PR DESCRIPTION
**Pull request by jmig5776**
#### Brief description of what is fixed or changed
This PR will define imageset union function for linear expression
Example:
```python
>>> img1 = ImageSet(Lambda(n, 4*n + 4), S.Integers)
>>> img2 = ImageSet(Lambda(n, 4*n), S.Integers)
>>> Union(img1, img2)
ImageSet(Lambda(n, 4*n), S.Integers)

>>> img1 = ImageSet(Lambda(n, n + 5), S.Integers)
>>> img2 = ImageSet(Lambda(n, 3*n), S.Integers)
>>> Union(img1, img2)
S.Integers

>>> img1 = ImageSet(Lambda(n, 2 * n * pi + pi / 4), S.Integers)
>>> img2 = ImageSet(Lambda(n, 2 * n * pi + 5 * pi / 4), S.Integers)
>>> Union(img1, img2)
ImageSet(Lambda(n, n * pi + pi / 4), S.Integers)
```

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by jmig5776** [comment]
ping @smichr @aktech @Yathartha22 @Shekharrajak @oscargus

**Comment by oscarbenjamin** [comment]
Both of the examples you showed might be handled better in general if ImageSet evaluated linear maps to a canonical form. Currently we have
```
In [1]: ImageSet(Lambda(n, 4*n + 4), S.Integers)                                                                                  
Out[1]: {4⋅n + 4 | n ∊ ℤ}
```
which could just simplify to `4*n` and
```
In [2]: ImageSet(Lambda(n, n + 5), S.Integers)                                                                                    
Out[2]: {n + 5 | n ∊ ℤ}
```
which could simplify to ℤ anyway.

**Comment by jmig5776** [comment]
This PR is basically to get all of your opinions on the procedure to unify two `ImageSet`. The algorithm that is implemented here is most fundamental. Please bear with me

We have to unify two sets i.e `a*n + b` and `c*n + d` . So which means if both of these expressions can be written in a single `x*n + y`.
So what I do is I maintain a `list1=[]`, `p = lcm(a,c)` in these following steps
1.  I insert all values of a*n +b for n = 0 to up to a*n + b < 2*p.
2. In the same list I Insert all c*n +d for n = 0 to up to c*n + d < 2*p.
3. Now from the list from first and second element I find the expression x*n + y which satisfies the pattern.
4. I check if this expression gives correct values on the whole `list1`. If yes then it is the required expression otherwise it's not.

Please consider this point while reviewing the PR. This PR is meant to get a proper procedure first for defining imageset union and not the code. The code can be improved later on.

**Comment by jmig5776** [comment]
@oscarbenjamin Please go through the tests in files changed. These were just basic examples. I will edit the PR description with complex examples.

**Comment by oscarbenjamin** [comment]
@jmig5776 I realise that this PR handles more complicated cases as well. I just wanted to remark that related simplifications can happen in `ImageSet.eval`.

**Comment by oscargus** [comment]
Look at #16033 and even more relevant the older PRs mentioned there. (Too tired at the moment to make a more relevant comment, but there are some good discussions there.)

**Comment by jmig5776** [comment]
Yeah, @oscargus I went through all the PRs before. @Shekharrajak I would like to have your opinion here for the method involved here. @smichr Can you please guide me here how I can reduce code here if this method seems convinient to you.

**Comment by jmig5776** [comment]
Some more improvement in code and documentation will be done in next commits.

**Comment by jmig5776** [comment]
@aktech @Yathartha22 @Shekharrajak Please review this PR.

File: sympy/solvers/tests/test_solveset.py [hunk 1]
```diff
@@ -1238,13 +1236,13 @@ def test_trig_system():
...
-    soln1 = (ImageSet(Lambda(n, 2*n*pi + pi/2), S.Integers),)
-    soln = FiniteSet(soln1)
-    assert nonlinsolve([sin(x) - 1, cos(x)], x) == soln
 
 
 @XFAIL
 def test_trig_system_fail():
+    soln1 = (ImageSet(Lambda(n, 2*n*pi + pi/2), S.Integers),)
+    soln = FiniteSet(soln1)
+    assert nonlinsolve([sin(x) - 1, cos(x)], x) == soln
```
Comment (jmig5776) [inline sympy/solvers/tests/test_solveset.py]: I dont know why this test is giving wrong solution but i debugged it with python debugger, the expression entering in ImageSet Union are giving correct result. Something else is giving wrong answer. I will try to debug it. Rest all of the tests are giving correct result.

**Comment by Yathartha22** [comment]
Will look at it in a day or so. You can remove the [WIP] if this is ready for review.

**Comment by jmig5776** [comment]
Mentors you can review this PR.

**Comment by jmig5776** [comment]
@aktech @Yathartha22 @Shekharrajak  Please review this PR also.

**Comment by czgdp1807** [comment]
Closing in favour of https://github.com/sympy/sympy/pull/17838

**Pull request closed without merging**